Rubberduck.Core/AutoComplete/Service/AutoCompleteService.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            _configService = configService;
+            _configService.SettingsChanged += ConfigServiceSettingsChanged;
+
+            _handlers = handlers;
+            InitializeConfig(); // todo: remove I/O work from constructor
+        }
+
+        private bool _initializing;
+
+        private void InitializeConfig()
+        {
+            _initializing = true;
+            // No reason to think this would throw, but if it does, _initializing state needs to be reset.
+            try
+            {
+                if (!_initialized)
+                {
+                    var config = _configService.LoadConfiguration();
+                    ApplyAutoCompleteSettings(config);
+                }
+            }
+            finally
+            {
+                _initializing = false;
+            }
+        }
+
+        public void Enable()
+        {
+            if (!_initializing)
+            {
+                InitializeConfig();
+            }
+
+            if (!_enabled)
+            {
+                VBENativeServices.KeyDown += HandleKeyDown;
+                VBENativeServices.IntelliSenseChanged += HandleIntelliSenseChanged;
+                _enabled = true;
+            }
+        }
+
+        public void Disable()
+        {
+            if (_enabled && _initialized)
+            {
+                VBENativeServices.KeyDown -= HandleKeyDown;
+                VBENativeServices.IntelliSenseChanged -= HandleIntelliSenseChanged;
+                _enabled = false;
+            }
+        }
+
+        private void HandleIntelliSenseChanged(object sender, IntelliSenseEventArgs e)
+        {
+            _popupShown = e.Visible;
+        }
+
+        private void ConfigServiceSettingsChanged(object sender, ConfigurationChangedEventArgs e)
+        {
+            var config = _configService.LoadConfiguration();
+            ApplyAutoCompleteSettings(config);
+        }
+        
+        public void ApplyAutoCompleteSettings(Configuration config)
+        {
+            _settings = config.UserSettings.AutoCompleteSettings;
+            if (_settings.IsEnabled)
+            {
+                Enable();
+            }
+            else
+            {
+                Disable();
+            }
+            _initialized = true;
+        }
+
+        private bool WillHandle(AutoCompleteEventArgs e)
+        {
+            Debug.Assert(_settings != null);
+
+            if (!_enabled)
+            {
+                Debug.Assert(_enabled, "KeyDown controller is executing, but auto-completion service is disabled.");
+                return false; // release build should bail out here!
+            }
+
+            if (_popupShown || e.Character == default && e.IsDeleteKey)
+            {
+                return false;
+            }
+
+            var module = e.Module;
+            using (var pane = module.CodePane)
```

Comment:
Why explicitly extract the event's module into a local var? Are we worried that someone else might change the Module while we handle the event? If yes: why? We only use it to access the CodePane of the Module...

Reply:
Artifact. Wasn't sure if I should be `using` the module.

Reply:
@Vogel612  actually... not extracting it fires a COM wrapper warning.